### PR TITLE
Add basic relation

### PR DESCRIPTION
**Pull request by simkim**

### CODE REVIEW

File: lib/couchbase-orm/relation.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+            def where(**conds)
+                CouchbaseOrm_Relation.new(**initializer_arguments.merge(where: conds))
+            end
+
+            def order(**order)
+                CouchbaseOrm_Relation.new(**initializer_arguments.merge(order: order))
+            end
+
+            def r_all
```
Comment (pimpin) [inline lib/couchbase-orm/relation.rb:53]: Why this name an not a simple `all` ?

File: lib/couchbase-orm/relation.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+            alias :to_a :to_ary
+
+            def delete_all
+                ids = query.to_a
+                CouchbaseOrm::Connection.bucket.default_collection.remove_multi(ids) unless ids.empty?
+            end
+
+            def where(**conds)
+                CouchbaseOrm_Relation.new(**initializer_arguments.merge(where: conds))
```
[the remark this answers is not in the document]
Comment (simkim) [reply lib/couchbase-orm/relation.rb:46]: Yes, on the todo

File: lib/couchbase-orm/relation.rb [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                @order = merge_order(**order) if order
+                @where = merge_where(where, _not) if where
+                CouchbaseOrm::logger.debug "- #{to_s}"
+            end
+
+            def to_s
+                "CouchbaseOrm_Relation: #{@model} where:#{@where.inspect} order:#{@order.inspect} limit: #{@limit}"
+            end
+
+            def build_order
```
Comment (giallon) [inline lib/couchbase-orm/relation.rb:21]: pass to private all build methods

File: lib/couchbase-orm/relation.rb [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+            delegate :each, :map, :collect, :to => :to_ary
+
+            def delete_all
+                CouchbaseOrm::logger.debug("Delete all: #{self}")
+                ids = query.to_a
+                CouchbaseOrm::Connection.bucket.default_collection.remove_multi(ids) unless ids.empty?
+            end
+            
+            def merge_where(conds, _not = false)
```
Comment (giallon) [inline lib/couchbase-orm/relation.rb:78]: pass to private

File: lib/couchbase-orm/relation.rb [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+            def where(**conds)
+                CouchbaseOrm_Relation.new(**initializer_arguments.merge(where: merge_where(conds)))
+            end
+
+            def not(**conds)
+                CouchbaseOrm_Relation.new(**initializer_arguments.merge(where: merge_where(conds, _not: true)))
+            end
+
+            def merge_order(*lorder, **horder)
```
Comment (giallon) [inline lib/couchbase-orm/relation.rb:90]: pass to private

File: lib/couchbase-orm/relation.rb [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            def to_ary
+                query.results { |res| @model.find(res) }.to_ary
+            end
+
+            alias :to_a :to_ary
+
+            delegate :each, :map, :collect, :to => :to_ary
+
+            def delete_all
+                CouchbaseOrm::logger.debug("Delete all: #{self}")
```
Comment (giallon) [inline lib/couchbase-orm/relation.rb:73]: pass debug message to methods

**Review by giallon — CHANGES_REQUESTED** [review]

File: lib/couchbase-orm/relation.rb [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                @order = merge_order(**order) if order
+                @where = merge_where(where, _not) if where
+                CouchbaseOrm::logger.debug "- #{to_s}"
+            end
+
+            def to_s
+                "CouchbaseOrm_Relation: #{@model} where:#{@where.inspect} order:#{@order.inspect} limit: #{@limit}"
+            end
+
+            def build_limit
```
Comment (giallon) [inline lib/couchbase-orm/relation.rb:21]: pass build_limit to private

File: lib/couchbase-orm/relation.rb [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+            def limit(limit)
+                CouchbaseOrm_Relation.new(**initializer_arguments.merge(limit: limit))
+            end
+
+            def all
+                CouchbaseOrm_Relation.new(**initializer_arguments)
+            end
+
+            def initializer_arguments
```
Comment (giallon) [inline lib/couchbase-orm/relation.rb:83]: pass initializer_arguments to private

**Review by giallon — CHANGES_REQUESTED** [review]

**Review by giallon — APPROVED** [review]

**Pull request merged**